File: typescript/src/domain/entities/Post.ts
import type { Uuid } from '../../application/wrapper/UuidWrapper'
import type { GetAttributes } from '../../helpers/TypesystemHelper'
import type { CreatePost } from '../commands/CreatePost'

export type PostId = Uuid

const inappropriateWords = ['orange', 'elephant', 'ice cream']

export class Post {
  private constructor(
    public readonly postId: PostId,
    public readonly text: string,
    public readonly dateTime: string
  ) {}

  private isValid(): boolean {
    const lowerText = this.text.toLowerCase()
    return (
      this.text.length > 0 &&
      !inappropriateWords.some(word => lowerText.includes(word))
    )
  }

  static create(postId: PostId, createPost: CreatePost): Post {
    const post = new Post(postId, createPost.text, new Date().toISOString())
    if (!post.isValid()) {
      throw new InvalidPost()
    }
    return post
  }

  static createFromAttributes(attributes: PostAttributes): Post {
    return new Post(attributes.postId, attributes.text, attributes.dateTime)
  }
}

export type PostAttributes = GetAttributes<Post>

export class InvalidPost extends Error {}

File: scala/src/main/scala/openchat/domain/entities/Post.scala
package openchat.domain.entities

import openchat.domain.commands.CreatePost
import openchat.domain.entities.Post.inappropriateWords
import openchat.domain.errors.InvalidPost

import java.time.format.DateTimeFormatter
import java.time.{Clock, Instant, ZoneOffset}

final case class Post(postId: String, text: String, dateTime: String) {
  private def isValid: Boolean = {
    val lower = text.toLowerCase
    text.nonEmpty && !inappropriateWords.exists(w => lower.contains(w))
  }
}

class InvalidPostException extends RuntimeException

object Post {
  def create(postId: String, cmd: CreatePost, clock: Clock = Clock.systemUTC()): Either[InvalidPost.type, Post] = {
    val post = Post(postId, cmd.text, dateFormatter.format(Instant.now(clock)))
    if (!post.isValid) Left(InvalidPost)
    else Right(post)
  }

  private val inappropriateWords: List[String] = List("orange", "elephant", "ice cream")

  private val dateFormatter: DateTimeFormatter =
    DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC)
}
